streamer/m3u8_concater.py

```diff
@@ -0,0 +1,957 @@
+              codec_division[codec][min(i, len(codec_division[codec]) - 1)])
+    
+    concat_vid_playlists: List[MediaPlaylist] = []
+    for codec, resolution_division in division.items():
+      for resolution, vid_playlists in resolution_division.items():
+        stream_info: Dict[str, str] = MediaPlaylist._similar_stream_info(
+        vid_playlists)
+        # Get a unique URI.
+        # NOTE: stream variants don't have a NAME attribute.
+        _, stream_info['URI'] = MediaPlaylist._next_unq_name()
+        # Get the peak and average bandwidth for this codec-resolution pair.
+        (stream_info['BANDWIDTH'],
+         stream_info['AVERAGE-BANDWIDTH']) = MediaPlaylist._get_bandwidth(
+            vid_playlists, durations)
+        # Get all the codecs that will be inside the new variant stream playlist.
+        stream_info['CODECS'] = MediaPlaylist._get_codec(vid_playlists)
+        concat_vid_playlist = MediaPlaylist(stream_info)
+        concat_vid_playlist.target_duration = MediaPlaylist._max_targer_dur(
+            vid_playlists)
+        for vid_playlist in vid_playlists:
+          concat_vid_playlist.content += vid_playlist.content
+          # Add a discontinuity after each period.
+          concat_vid_playlist.content += '#EXT-X-DISCONTINUITY\n\n'
+        concat_vid_playlists.append(concat_vid_playlist)
+    
+    return concat_vid_playlists
+
+
+class MasterPlaylist:
+  """A class representing a master playlist."""
+  
+  header = ''
+  """A common header for MasterPlaylist objects."""
+  
+  def __init__(self,
+               file_name: Optional[str] = None,
+               output_dir: Optional[str] = None,
+               packager: Optional[PackagerNode] = None):
+    """Given the path to the master playlist file, this method will read that file
+    and search for MediaPlaylists inside it, and instantiate a MediaPlaylist object
+    for each one found.
+    
+    The (MediaPlaylist)'s __init__ method will parse each media playlist and change the
+    path of media segments to be relative to `output_dir`.
+    
+    if `file_name` is None, an empty MasterPlaylist object is returned.
+    """
+    
+    self.playlists: List[MediaPlaylist] = []
+    self.duration = 0.0
+    
+    if file_name is None:
+      # Do not read. The media playlists will be appended manually.
+      return
+
+    # If there is a file to read, we MUST have a bound packager node
+    # to match its streams with each media playlist.
+    assert packager is not None
+    
+    # We will fill this map, file names to (OutputStream)s and will give it out
+    # to each media playlist we initialize.  Each media playlist then will pick
+    # the right output stream for itself.
+    streams_map: Dict[str, OutputStream] = {}
+    
+    # Add a mapping between single segment file names and their output stream.
+    streams_map.update({
+        outstream
+        .get_single_seg_file()
+        .write_end(): outstream
```

Comment:
Will these fit on one line?  I think I would find it easier to read that way.

Reply:
This does.